### DISCUSSION TITLE
Release 5.0.0

### DISCUSSION THREAD

**Discussion by armandino** [instancio/instancio#1124] [Announcements]
#### Renamed classes:

- `org.instancio.InstancioOfClassApi` to `InstancioClassApi`
- `org.instancio.InstancioOfCollectionApi` to `InstancioCollectionsApi`

#### 2. Replaced `org.instancio.Gen` with `org.instancio.Instancio.gen()`

```java
// Before
import org.instancio.Gen;

int n = Gen.ints().range(1, 10).get();

// Now
import static org.instancio.Instancio.gen;

int n = gen().ints().range(1, 10).get();
```

#### 3. Removed method `org.instancio.generator.ValueSpec.toModel()`:

```java
Model<Integer> model = gen().ints().range(1, 10)
        .toModel(); // toModel() method no longer available
```

The above can still be achieved using `Instancio.of()` API:

```java
Model<Integer> model = Instancio.of(Integer.class)
        .generate(Select.root(), gen -> gen.ints().range(1, 10))
        .toModel();
```

#### `Feed` API for generating data from an external resource (CSV or JSON)

**Example data file: `persons.csv`**

```csv
firstName,lastName,age
John,Doe,21
Alice,Smith,34
# ...
```

```java
List<Person> person = Instancio.ofList(Person.class)
    .applyFeed(all(Person.class), feed -> feed.ofResource("persons.csv"))
    .create();

// Output:
// [Person[firstName="John", lastName="Doe", age=21],
//  Person[firstName="Alice", lastName="Smith", age=34],
//  ...]
```

A feed can also be defined as an interface, which offers additional features and allows
the feed to be reused across multiple tests.

```java
@Feed.Source(name = "persons.csv")
interface PersonFeed extends Feed {

  FeedSpec<String> firstName(); // maps to 'firstName' property in the data file

  @DataSpec("lastName") // maps to `lastName` property in the data file
  FeedSpec<String> surname();

  @TemplateSpec("${firstName} ${lastName}")
  FeedSpec<String> fullName();

  @FunctionSpec(params = "age", provider = DateOfBirthCalculator.class)
  FeedSpec<LocalDate> dateOfBirth();

  class DateOfBirthCalculator implements FunctionProvider {
    LocalDate toDateOfBirth(int age) {
      return LocalDate.now().minusYears(age);
    }
  }
}

// Sample usage
List<Person> person = Instancio.ofList(Person.class)
    .applyFeed(all(Person.class), feed -> feed.of(PersonFeed.class))
    .create();

// Output:
// [Person[firstName="John", surname="Doe", fullName="John Doe", age=21, ...],
//  ...]
```

More info: https://www.instancio.org/user-guide/#data-feeds

#### `@Given` annotation for injecting test fields and parameters

```java
@ExtendWith(InstancioExtension.class)
class ExampleTest {

    @Given
    private String randomStringField;

    @Test
    void example1(@Given List<String> randomList) {
        assertThat(randomStringField).isNotBlank();
        assertThat(randomList).isNotEmpty();
    }

    @ValueSource(strings = {"foo", "bar"})
    @ParameterizedTest
    void example2(String fooOrBar, @Given long randomLong) {
        assertThat(fooOrBar).isIn("foo", "bar");
        assertThat(randomLong).isPositive();
    }
}
```

The `@Given` annotation can also be used with a custom `GivenProvider` implementation, e.g.

```java
    @Given(CustomStringProvider.class)
    private String customString;
```

Additionally, the `@Given` annotation can be used to define custom annotations:

```java
@Given(NumericStringProvider.class)
@Target({ElementType.FIELD, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
@interface NumericString {

    int length();

    class NumericStringProvider implements GivenProvider {
        @Override
        public Object provide(ElementContext context) {
            NumericString numericString = context.getAnnotation(NumericString.class);
            return context.random().digits(numericString.length());
        }
    }
}

// Sample usage:
@ExtendWith(InstancioExtension.class)
class ExampleTest {

    @Test
    void example(@NumericString(length = 5) String digits) {
        assertThat(digits).hasSize(5).containsOnlyDigits();
    }
 }
```

More info: https://www.instancio.org/user-guide/#given-injection

#### `@InstancioSource` enhancements

- Added `samples` attribute - specifies the number of times a test method annotated with `@InstancioSource` will run.
- Added `Keys.INSTANCIO_SOURCE_SAMPLES` setting - configures the default number of samples.

By default, starting from version `5.0.0`, tests annotated with `@InstancioSource` will run 100 times.  This is a potentially breaking change, as previously, they were executed only once.

**Example:**

```java
@ExtendWith(InstancioExtension.class)
class ExampleTest {

    @ParameterizedTest
    @InstancioSource
    void example1(Foo foo) {
        // will run 100 times with different instances of Foo
    }

    @ParameterizedTest
    @InstancioSource(samples = 250)
    void multipleArguments(Bar bar, Baz baz) {
        // will run 250 times with different instances of Bar and Baz
    }
}
```

#### Support for generating intervals of various types, such as numeric and temporal values

Example: generate a list of vacations with start/end dates.

```java
record Vacation(LocalDate start, LocalDate end) {}
```

- The first vacation should start on `2000-01-01`.
- Each vacation should last between 1-3 weeks.
- There should be 12-15 months between vacations.

```java
IntervalSupplier<LocalDate> vacationDates = Instancio.gen()
    .intervalStarting(LocalDate.of(2000, 1, 1))
    .nextStart((end, random) -> end.plusMonths(random.intRange(12, 15)))
    .nextEnd((start, random) -> start.plusWeeks(random.intRange(1, 3)))
    .get();

List<Vacation> vacations = Instancio.ofList(Vacation.class)
    .size(4)
    .supply(field(Vacation::start), vacationDates.start())
    .supply(field(Vacation::end), vacationDates.end())
    .create();

 // Sample output:
 // [Vacation[start=2000-01-01, end=2000-01-15],
 //  Vacation[start=2001-02-15, end=2001-03-08],
 //  Vacation[start=2002-03-08, end=2002-03-15],
 //  Vacation[start=2003-06-15, end=2003-07-06]]
```

#### Support for `assign().valueOf()` value with randomisation

Example: create a time period between 1 and 30 days, inclusive.

```java
record TimePeriod(LocalDate start, LocalDate end) {}

TimePeriod period = Instancio.of(TimePeriod.class)
    .assign(Assign.valueOf(TimePeriod::start)
            .to(TimePeriod::end)
            .as((LocalDate start, Random random) -> start.plusDays(random.intRange(1, 30))))
    .create();

// Sample output: TimePeriod[start=2012-01-22, end=2012-01-30]
```

#### Support for Hibernate Validator constraints

- Brazil: `@CNPJ`, `@CPF` `@TituloEleitoral` (@brunodmartins)
- Russia: `@INN` (@EvaristeGalois11)

#### Value spec for shuffling arrays and collections

-----
* Removed deprecated `instancio-quickcheck` module by @armandino in https://github.com/instancio/instancio/pull/1063
* Breaking change: removed `ValueSpec.toModel()` method by @armandino in https://github.com/instancio/instancio/pull/1064
* Breaking changes: removed (experimental) `org.instancio.Gen` class by @armandino in https://github.com/instancio/instancio/pull/1065
* Minor refactoring of API interfaces by @armandino in https://github.com/instancio/instancio/pull/1066
* Support for assigning value with randomisation by @armandino in https://github.com/instancio/instancio/pull/1067
* Bump version.jooq from 3.19.9 to 3.19.10 by @dependabot in https://github.com/instancio/instancio/pull/1070
* Bump org.apache.maven.plugins:maven-jar-plugin from 3.4.1 to 3.4.2 by @dependabot in https://github.com/instancio/instancio/pull/1073
* Bump com.github.spotbugs:spotbugs-maven-plugin from 4.8.5.0 to 4.8.6.0 by @dependabot in https://github.com/instancio/instancio/pull/1076
* Bump org.apache.maven.plugins:maven-failsafe-plugin from 3.2.5 to 3.3.0 by @dependabot in https://github.com/instancio/instancio/pull/1071
* Bump org.apache.maven.plugins:maven-surefire-plugin from 3.2.5 to 3.3.0 by @dependabot in https://github.com/instancio/instancio/pull/1069
* Bump com.github.spotbugs:spotbugs-maven-plugin from 4.8.6.0 to 4.8.6.1 by @dependabot in https://github.com/instancio/instancio/pull/1081
* Bump org.projectlombok:lombok from 1.18.32 to 1.18.34 by @dependabot in https://github.com/instancio/instancio/pull/1080
* Bump org.junit:junit-bom from 5.10.2 to 5.10.3 by @dependabot in https://github.com/instancio/instancio/pull/1082
* Support for Hibernate Validator `@CPF` annotation (Brazil) #1006  by @brunodmartins in https://github.com/instancio/instancio/pull/1083
* Bump org.codehaus.mojo:versions-maven-plugin from 2.16.2 to 2.17.0 by @dependabot in https://github.com/instancio/instancio/pull/1084
* Brazil CNPJ generator by @brunodmartins in https://github.com/instancio/instancio/pull/1085
* Add INN Generator by @EvaristeGalois11 in https://github.com/instancio/instancio/pull/1086
* Fix timestamp generator test in CET timezone by @EvaristeGalois11 in https://github.com/instancio/instancio/pull/1087
* Bump org.codehaus.mojo:animal-sniffer-maven-plugin from 1.23 to 1.24 by @dependabot in https://github.com/instancio/instancio/pull/1089
* Remove asm dependency from animal sniffer plugin by @EvaristeGalois11 in https://github.com/instancio/instancio/pull/1090
* Update maven wrapper and switch to type only-script by @EvaristeGalois11 in https://github.com/instancio/instancio/pull/1091
* Brazil Titulo Eleitoral generator by @brunodmartins in https://github.com/instancio/instancio/pull/1092
* Bump com.github.spotbugs:spotbugs-maven-plugin from 4.8.6.1 to 4.8.6.2 by @dependabot in https://github.com/instancio/instancio/pull/1093
* Bump org.apache.maven.plugins:maven-surefire-plugin from 3.3.0 to 3.3.1 by @dependabot in https://github.com/instancio/instancio/pull/1096
* Bump org.apache.maven.plugins:maven-failsafe-plugin from 3.3.0 to 3.3.1 by @dependabot in https://github.com/instancio/instancio/pull/1095
* New Feed API for generating data from external sources by @armandino in https://github.com/instancio/instancio/pull/1078
* Bump com.fasterxml.jackson.core:jackson-databind from 2.17.1 to 2.17.2 by @dependabot in https://github.com/instancio/instancio/pull/1097
* Log seed at TRACE level (#1075) by @armandino in https://github.com/instancio/instancio/pull/1098
* Value spec for shuffling arrays and collections by @armandino in https://github.com/instancio/instancio/pull/1099
* Add `samples` attribute to `InstancioSource` to support repeated tests by @armandino in https://github.com/instancio/instancio/pull/1100
* Bump org.codehaus.mojo:versions-maven-plugin from 2.17.0 to 2.17.1 by @dependabot in https://github.com/instancio/instancio/pull/1104
* Bump org.apache.commons:commons-lang3 from 3.14.0 to 3.15.0 by @dependabot in https://github.com/instancio/instancio/pull/1108
* `instancio-junit`: `@Given` annotation for wiring fields and parameters (#1072) by @armandino in https://github.com/instancio/instancio/pull/1107
* Document temporal spec `past()` and `future()` as non-deterministic (#1105) by @armandino in https://github.com/instancio/instancio/pull/1110
* Experimental API breaking change: renamed CSV spec `separator()` to `delimiter()` by @armandino in https://github.com/instancio/instancio/pull/1111
* New setting key: `Keys.MAX_GENERATION_ATTEMPTS` by @armandino in https://github.com/instancio/instancio/pull/1112
* Upgrade AssertJ to 3.26.3 by @armandino in https://github.com/instancio/instancio/pull/1113
* Improved max generation attempts exceeded error message by @armandino in https://github.com/instancio/instancio/pull/1114
* Improvements to unused selectors error message by @armandino in https://github.com/instancio/instancio/pull/1115
* Breaking change: renamed `CartesianProductApi` class (experimental API) by @armandino in https://github.com/instancio/instancio/pull/1116
* Breaking change: renamed `InstancioOfClassApi` by @armandino in https://github.com/instancio/instancio/pull/1117
* Breaking change: renamed `InstancioOfCollectionApi` by @armandino in https://github.com/instancio/instancio/pull/1118
* Bump org.apache.maven.plugins:maven-javadoc-plugin from 3.7.0 to 3.8.0 by @dependabot in https://github.com/instancio/instancio/pull/1120
* Bump com.github.siom79.japicmp:japicmp-maven-plugin from 0.21.2 to 0.22.0 by @dependabot in https://github.com/instancio/instancio/pull/1119
* Support for generating intervals by @armandino in https://github.com/instancio/instancio/pull/1121
* Cleanup `ValueSpec` by @armandino in https://github.com/instancio/instancio/pull/1122
* Bug fix: error using `setModel()` with root selector by @armandino in https://github.com/instancio/instancio/pull/1123

## New Contributors
* @brunodmartins made their first contribution in https://github.com/instancio/instancio/pull/1083

**Full Changelog**: https://github.com/instancio/instancio/compare/instancio-parent-4.8.0...instancio-parent-5.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/instancio/instancio/releases/tag/instancio-parent-5.0.0'>5.0.0</a>.</em>

**Comment by nandorholozsnyak** [comment]
Huge update. Congrats.

Quicky question on the breaking changes @armandino , have you considered to provide (in the future or now?) [OpenRewrite Recipes](https://docs.openrewrite.org/concepts-explanations/recipes)?

**Reply by armandino** [reply]
Thanks @nandorholozsnyak! I can look into it, but it might take me some time as I haven't had a chance to use OpenRewrite yet. If anyone is interested in taking this on, that will likely be much faster :)
